### PR TITLE
Move Days-Elapsed logic up to the job level

### PR DESCRIPTION
**Pull request by ed0906**

### CODE REVIEW

**Comment by ed0906** [comment]
I'm happy for this not to be merged if you prefer it as it is. Thought this would improve the logging a bit.

File: app/uk/gov/hmrc/bindingtariffclassification/scheduler/DaysElapsedJob.scala [hunk 1]
```diff
@@ -40,8 +41,27 @@ class DaysElapsedJob @Inject()(appConfig: AppConfig, caseService: CaseService) e
   override def interval: FiniteDuration = FiniteDuration(jobConfig.intervalDays, TimeUnit.DAYS)
 
   override def execute(): Future[String] = {
-    caseService.incrementDaysElapsedIfAppropriate(jobConfig.intervalDays, appConfig.clock)
-      .map(count => s"Incremented the Days Elapsed for [$count] cases.")
+    ifTodayIsAWorkingDay { () =>
+      caseService.incrementDaysElapsed(jobConfig.intervalDays)
+        .map(count => s"Incremented the Days Elapsed for [$count] cases.")
```
Comment (gokyo) [inline app/uk/gov/hmrc/bindingtariffclassification/scheduler/DaysElapsedJob.scala]: you can use `_`

File: app/uk/gov/hmrc/bindingtariffclassification/scheduler/DaysElapsedJob.scala [hunk 2]
```diff
@@ -29,7 +30,7 @@ import scala.concurrent.Future
 import scala.concurrent.duration.FiniteDuration
 
 @Singleton
-class DaysElapsedJob @Inject()(appConfig: AppConfig, caseService: CaseService) extends ScheduledJob {
+class DaysElapsedJob @Inject()(appConfig: AppConfig, caseService: CaseService, bankHolidaysConnector: BankHolidaysConnector) extends ScheduledJob {
```
Comment (gokyo) [inline app/uk/gov/hmrc/bindingtariffclassification/scheduler/DaysElapsedJob.scala]: double check that `appConfig.daysElapsed` reads the correct full path from application.conf

File: app/uk/gov/hmrc/bindingtariffclassification/scheduler/DaysElapsedJob.scala [hunk 3]
```diff
@@ -40,8 +41,27 @@ class DaysElapsedJob @Inject()(appConfig: AppConfig, caseService: CaseService) e
...
   override def execute(): Future[String] = {
-    caseService.incrementDaysElapsedIfAppropriate(jobConfig.intervalDays, appConfig.clock)
-      .map(count => s"Incremented the Days Elapsed for [$count] cases.")
+    ifTodayIsAWorkingDay { () =>
+      caseService.incrementDaysElapsed(jobConfig.intervalDays)
+        .map(count => s"Incremented the Days Elapsed for [$count] cases.")
+    }
+  }
+
+  private def ifTodayIsAWorkingDay(function: () => Future[String]): Future[String] = {
```
Comment (gokyo) [inline app/uk/gov/hmrc/bindingtariffclassification/scheduler/DaysElapsedJob.scala]: I would return `Future[Boolean]`.
Then, where this method is called, I would do a `.map` and execute a block of code (your `function`) if it is `true`.

You can use `Logger.info` for the messages.
Also, it is a good idea to log in Kibana a message when the task started and a message when it finishes.
In the Platform Readiness Assessment it is a requirement for each scheduled job.

Comment (ed0906) [reply app/uk/gov/hmrc/bindingtariffclassification/scheduler/DaysElapsedJob.scala]: I agree, will change

File: app/uk/gov/hmrc/bindingtariffclassification/scheduler/DaysElapsedJob.scala [hunk 4]
```diff
@@ -39,9 +41,30 @@ class DaysElapsedJob @Inject()(appConfig: AppConfig, caseService: CaseService) e
...
+        .map(count => Logger.info(s"Scheduled Job [$name]: Incremented the Days Elapsed for [$count] cases."))
+    }
+  }
+
+  private def ifTodayIsAWorkingDay(function: => Future[Unit]): Future[Unit] = {
+    val today = LocalDate.now(appConfig.clock)
+    val dayOfTheWeek = today.getDayOfWeek
+    if (dayOfTheWeek == DayOfWeek.SATURDAY || dayOfTheWeek == DayOfWeek.SUNDAY) {
+      Logger.info(s"Scheduled Job [$name]: Skipping today as it is a Weekend")
+      Future.successful()
```
Comment (gokyo) [inline app/uk/gov/hmrc/bindingtariffclassification/scheduler/DaysElapsedJob.scala]: You need to add an extra `()` for the Unit
```
Future.successful(())
```

File: app/uk/gov/hmrc/bindingtariffclassification/scheduler/DaysElapsedJob.scala [hunk 5]
```diff
@@ -39,9 +41,30 @@ class DaysElapsedJob @Inject()(appConfig: AppConfig, caseService: CaseService) e
...
+    } else {
+      bankHolidaysConnector.get()
+        .map(_.contains(today))
+        .flatMap {
+          case true =>
+            Logger.info(s"Scheduled Job [$name]: Skipping today as it is a Bank Holiday")
+            Future.successful()
```
Comment (gokyo) [inline app/uk/gov/hmrc/bindingtariffclassification/scheduler/DaysElapsedJob.scala]: extra brackets here as well

File: app/uk/gov/hmrc/bindingtariffclassification/scheduler/DaysElapsedJob.scala (hunk 4 above)
Comment (ed0906) [reply app/uk/gov/hmrc/bindingtariffclassification/scheduler/DaysElapsedJob.scala]: Fixed

**Pull request merged**